README
NAME
    PYX::Hist - Processing PYX data or file and print histogram.

SYNOPSIS
     use PYX::Hist;

     my $obj = PYX::Hist->new(%parameters);
     $obj->parse($pyx, $out);
     $obj->parse_file($input_file, $out);
     $obj->parse_handler($input_file_handler, $out);

METHODS
  "new"
     my $obj = PYX::Hist->new(%parameters);

    Constructor.

    *       "output_handler"

             Output handler.
             Default value is \*STDOUT.

    Returns instance of object.

  "parse"
     $obj->parse($pyx, $out);

    Parse PYX text or array of PYX text and print histogram of PYX input. If
    $out not present, use 'output_handler'.

    Returns undef.

  "parse_file"
     $obj->parse_file($input_file, $out);

    Parse file with PYX data and print histogram of PYX input. If $out not
    present, use 'output_handler'.

    Returns undef.

  "parse_handler"
     $obj->parse_handler($input_file_handler, $out);

    Parse PYX handler and print histogram of PYX input. If $out not present,
    use 'output_handler'.

    Returns undef.

ERRORS
     new():
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

     parse():
             Bad end of element.
                     Element: %s
             Stack has some elements.

     parse_file():
             Bad end of element.
                     Element: %s
             Stack has some elements.

     parse_handler():
             Bad end of element.
                     Element: %s
             Stack has some elements.

EXAMPLE1
     use strict;
     use warnings;

     use PYX::Hist;

     # Example data.
     my $pyx = <<'END';
     (begin
     (middle
     (end
     -data
     )end
     (end
     -data
     )end
     )middle
     )begin
     END

     # PYX::Hist object.
     my $obj = PYX::Hist->new;

     # Parse.
     $obj->parse($pyx);

     # Output:
     # [ begin  ] 1
     # [ end    ] 2
     # [ middle ] 1

EXAMPLE2
     use strict;
     use warnings;

     use Error::Pure;
     use PYX::Hist;

     # Error output.
     $Error::Pure::TYPE = 'PrintVar';

     # Example data.
     my $pyx = <<'END';
     (begin
     (middle
     (end
     -data
     )middle
     )begin
     END

     # PYX::Hist object.
     my $obj = PYX::Hist->new;

     # Parse.
     $obj->parse($pyx);

     # Output:
     # PYX::Hist: Bad end of element.
     # Element: middle

EXAMPLE3
     use strict;
     use warnings;

     use Error::Pure;
     use PYX::Hist;

     # Error output.
     $Error::Pure::TYPE = 'PrintVar';

     # Example data.
     my $pyx = <<'END';
     (begin
     (middle
     (end
     -data
     )end
     )middle
     END

     # PYX::Hist object.
     my $obj = PYX::Hist->new;

     # Parse.
     $obj->parse($pyx);

     # Output:
     # PYX::Hist: Stack has some elements.

DEPENDENCIES
    Class::Utils, Error::Pure, List::Util, PYX::Parser.

SEE ALSO
    Task::PYX
        Install the PYX modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/PYX-Hist>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.09



Changes
0.09

0.08 2023-10-19T10:19:44+02:00
 - Fix typo in doc.
 - Improve constructor doc.
 - Update Module::Install to 1.21 version.

0.07 2023-03-30T15:48:16+02:00
 - Move bugtracker to github.
 - Regen META.yml file.

0.06 2023-03-30T12:53:11+02:00
 - Add names for example files.
 - Rewrite doc to use =head2.
 - Rewrite usage in tests from IO::CaptureOutput to Capture::Tiny.
 - Update copyright years.

0.05 2020-04-24T15:24:08+02:00
 - Fix Makefile.PL fails with no '.' in @INC.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.04 2016-02-23T14:13:20+01:00
 - Add EXAMPLE3 section to doc.
 - Fix EXAMPLE2 section in doc.
 - Fix minimal version of List::Util to 1.41.
 - Update copyright years.

0.03 2015-10-13T15:14:26+02:00
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Update Module::Install to 1.16 version.
 - Rewrite SEE ALSO section in doc to proper Task.

0.02 2015-01-31T19:45:32+01:00
 - Add PYX::Stack to SEE ALSO section in doc.
 - Add examples/ directory.
 - Fix abstract.

0.01 2015-01-26T00:19:40+01:00
 - First version.


Hist.pm
package PYX::Hist;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use List::Util qw(reduce);
use PYX::Parser;

our $VERSION = 0.09;

# Constructor.
sub new {
	my ($class, @params) = @_;
	my $self = bless {}, $class;

	# Output handler.
	$self->{'output_handler'} = \*STDOUT;

	# Process params.
	set_params($self, @params);

	# PYX::Parser object.
	$self->{'_pyx_parser'} = PYX::Parser->new(
		'callbacks' => {
			'end_element' => \&_end_element,
			'final' => \&_final,
			'start_element' => \&_start_element,
		},
		'non_parser_options' => {
			'hist' => {},
			'stack' => [],
		},
		'output_handler' => $self->{'output_handler'},
	);

	# Object.
	return $self;
}

# Parse pyx text or array of pyx text.
sub parse {
	my ($self, $pyx, $out) = @_;
	$self->{'_pyx_parser'}->parse($pyx, $out);
	return;
}

# Parse file with pyx text.
sub parse_file {
	my ($self, $file, $out) = @_;
	$self->{'_pyx_parser'}->parse_file($file, $out);
	return;
}

# Parse from handler.
sub parse_handler {
	my ($self, $input_file_handler, $out) = @_;
	$self->{'_pyx_parser'}->parse_handler($input_file_handler, $out);
	return;
}

# End of element.
sub _end_element {
	my ($pyx_parser_obj, $elem) = @_;
	my $stack_ar = $pyx_parser_obj->{'non_parser_options'}->{'stack'};
	if ($stack_ar->[-1] eq $elem) {
		pop @{$stack_ar};
	} else {
		err 'Bad end of element.',
			'Element', $elem;
	}
	return;
}

# Finalize.
sub _final {
	my $pyx_parser_obj = shift;
	my $stack_ar = $pyx_parser_obj->{'non_parser_options'}->{'stack'};
	if (@{$stack_ar} > 0) {
		err 'Stack has some elements.';
	}
	my $hist_hr = $pyx_parser_obj->{'non_parser_options'}->{'hist'};
	my $max_len = length reduce { length($a) > length($b) ? $a : $b }
		keys %{$hist_hr};
	my $out = $pyx_parser_obj->{'output_handler'};
	foreach my $key (sort keys %{$hist_hr}) {
		printf {$out} "[ %-${max_len}s ] %s\n", $key, $hist_hr->{$key};
	}
	return;
}

# Start of element.
sub _start_element {
	my ($pyx_parser_obj, $elem) = @_;
	my $stack_ar = $pyx_parser_obj->{'non_parser_options'}->{'stack'};
	push @{$stack_ar}, $elem;
	if (! $pyx_parser_obj->{'non_parser_options'}->{'hist'}->{$elem}) {
		$pyx_parser_obj->{'non_parser_options'}->{'hist'}->{$elem} = 1;
	} else {
		$pyx_parser_obj->{'non_parser_options'}->{'hist'}->{$elem}++;
	}
	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

PYX::Hist - Processing PYX data or file and print histogram.

=head1 SYNOPSIS

 use PYX::Hist;

 my $obj = PYX::Hist->new(%parameters);
 $obj->parse($pyx, $out);
 $obj->parse_file($input_file, $out);
 $obj->parse_handler($input_file_handler, $out);

=head1 METHODS

=head2 C<new>

 my $obj = PYX::Hist->new(%parameters);

Constructor.

=over 8

=item * C<output_handler>

 Output handler.
 Default value is \*STDOUT.

=back

Returns instance of object.

=head2 C<parse>

 $obj->parse($pyx, $out);

Parse PYX text or array of PYX text and print histogram of PYX input.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_file>

 $obj->parse_file($input_file, $out);

Parse file with PYX data and print histogram of PYX input.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_handler>

 $obj->parse_handler($input_file_handler, $out);

Parse PYX handler and print histogram of PYX input.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

 parse():
         Bad end of element.
                 Element: %s
         Stack has some elements.

 parse_file():
         Bad end of element.
                 Element: %s
         Stack has some elements.

 parse_handler():
         Bad end of element.
                 Element: %s
         Stack has some elements.

=head1 EXAMPLE1

=for comment filename=hist.pl

 use strict;
 use warnings;

 use PYX::Hist;

 # Example data.
 my $pyx = <<'END';
 (begin
 (middle
 (end
 -data
 )end
 (end
 -data
 )end
 )middle
 )begin
 END

 # PYX::Hist object.
 my $obj = PYX::Hist->new;

 # Parse.
 $obj->parse($pyx);

 # Output:
 # [ begin  ] 1
 # [ end    ] 2
 # [ middle ] 1

=head1 EXAMPLE2

=for comment filename=hist_with_bad_end_element.pl

 use strict;
 use warnings;

 use Error::Pure;
 use PYX::Hist;

 # Error output.
 $Error::Pure::TYPE = 'PrintVar';

 # Example data.
 my $pyx = <<'END';
 (begin
 (middle
 (end
 -data
 )middle
 )begin
 END

 # PYX::Hist object.
 my $obj = PYX::Hist->new;

 # Parse.
 $obj->parse($pyx);

 # Output:
 # PYX::Hist: Bad end of element.
 # Element: middle

=head1 EXAMPLE3

=for comment filename=hist_with_missing_element.pl

 use strict;
 use warnings;

 use Error::Pure;
 use PYX::Hist;

 # Error output.
 $Error::Pure::TYPE = 'PrintVar';

 # Example data.
 my $pyx = <<'END';
 (begin
 (middle
 (end
 -data
 )end
 )middle
 END

 # PYX::Hist object.
 my $obj = PYX::Hist->new;

 # Parse.
 $obj->parse($pyx);

 # Output:
 # PYX::Hist: Stack has some elements.

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<List::Util>,
L<PYX::Parser>.

=head1 SEE ALSO

=over

=item L<Task::PYX>

Install the PYX modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/PYX-Hist>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.09

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Processing PYX data or file and print histogram.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.07;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'PYX-Hist';
readme_from 'Hist.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.05;
requires 'Error::Pure' => 0.16;
requires 'List::Util' => 1.41;
requires 'PYX::Parser' => 0.02;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/PYX-Hist/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/PYX-Hist';
resources 'repository' => 'git://github.com/michal-josef-spacek/PYX-Hist';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'File::Object' => 0.07;
test_requires 'Capture::Tiny' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.09';

# Run.
WriteAll();
